src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -195,6 +203,45 @@ mod test {
         }
     }
 
+    #[test]
+    fn test_reverse() {
+        let cases = vec![
+            (
+                Datum::Bytes(b"hello".to_vec()),
+                Datum::Bytes(b"olleh".to_vec()),
+            ),
+            (Datum::Bytes(b"".to_vec()), Datum::Bytes(b"".to_vec())),
+            (
+                Datum::Bytes("数据库".as_bytes().to_vec()),
+                Datum::Bytes("库据数".as_bytes().to_vec()),
+            ),
+            (
+                Datum::Bytes("忠犬ハチ公".as_bytes().to_vec()),
+                Datum::Bytes("公チハ犬忠".as_bytes().to_vec()),
+            ),
+            (
+                Datum::Bytes("あなたのことが好きです".as_bytes().to_vec()),
+                Datum::Bytes("すでき好がとこのたなあ".as_bytes().to_vec()),
+            ),
+            (
+                Datum::Bytes("Bayern München".as_bytes().to_vec()),
+                Datum::Bytes("nehcnüM nreyaB".as_bytes().to_vec()),
+            ),
+            (
+                Datum::Bytes("Η Αθηνά  ".as_bytes().to_vec()),
+                Datum::Bytes("  άνηθΑ Η".as_bytes().to_vec()),
+            ),
+            (Datum::Null, Datum::Null),
+        ];
+        let mut ctx = EvalContext::default();
+        for (arg, exp) in cases {
+            let op = scalar_func_expr(ScalarFuncSig::Reverse, &[datum_expr(arg)]);
```

Comment:
I think we need to do `op.mut_field_type().set_charset(...)`. Now only `utf-8` is supported, so set it to `utf-8` is ok.

Reply:
@hicqu In TiDB,  for binary string types such as  `BINARY`, `VARBINARY` and so on,  the charset and collate is set to `binary`. These settings are pushed down to the kv layer. If we panic on non-utf8 charsets,  we can't deal with these binary string types.  For example, type `BINARY`  has the type setting as follows:
```rust
         for (input, exp) in cases {
             let input = string_datum_expr_with_tp(
                 input,
                 VAR_STRING,
                 BINARY_FLAG,
                 -1,
                 CHARSET_BIN.to_owned(),
                 COLLATION_BIN_ID,
             );

```

and the implementation of builtin `UPPER` in TiDB is as follows:
```go
func (b *builtinUpperSig) evalString(row chunk.Row) (d string, isNull bool, err error) {
	d, isNull, err = b.args[0].EvalString(b.ctx, row)
	if isNull || err != nil {
		return d, isNull, errors.Trace(err)
	}

	if types.IsBinaryStr(b.args[0].GetType()) {
		return d, false, nil
	}

	return strings.ToUpper(d), false, nil
}

```